Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,505 @@
+        return false;
+    }
+
+    Vector<EGLAttrib> attributes = {
+        EGL_WIDTH, static_cast<EGLAttrib>(gbm_bo_get_width(m_bo)),
+        EGL_HEIGHT, static_cast<EGLAttrib>(gbm_bo_get_height(m_bo)),
+        EGL_LINUX_DRM_FOURCC_EXT, static_cast<EGLAttrib>(gbm_bo_get_format(m_bo))
+    };
+
+    std::array<EGLAttrib, 6> planeAttributes {
+        EGL_DMA_BUF_PLANE0_FD_EXT, static_cast<EGLAttrib>(dmaBufFD),
+        EGL_DMA_BUF_PLANE0_OFFSET_EXT, static_cast<EGLAttrib>(gbm_bo_get_offset(m_bo, 0)),
+        EGL_DMA_BUF_PLANE0_PITCH_EXT, static_cast<EGLAttrib>(gbm_bo_get_stride_for_plane(m_bo, 0))
+    };
+    attributes.append(std::span<const EGLAttrib> { planeAttributes });
+
+    if (m_modifier != DRM_FORMAT_MOD_INVALID) {
+        std::array<EGLAttrib, 4> modifierAttributes {
+            EGL_DMA_BUF_PLANE0_MODIFIER_HI_EXT, static_cast<EGLAttrib>(m_modifier >> 32),
+            EGL_DMA_BUF_PLANE0_MODIFIER_LO_EXT, static_cast<EGLAttrib>(m_modifier & 0xffffffff)
+        };
+        attributes.append(std::span<const EGLAttrib> { modifierAttributes });
+    }
+
+    attributes.append(EGL_NONE);
+
+    auto& display = WebCore::PlatformDisplay::sharedDisplay();
+    m_eglImage = display.createEGLImage(EGL_NO_CONTEXT, EGL_LINUX_DMA_BUF_EXT, nullptr, attributes);
+
+    close(dmaBufFD);
+    return !!m_eglImage;
+}
+
+bool MemoryMappedGPUBuffer::exportGBMBufferObjectAsDMABuf()
+{
+    if (auto dmaBufFD = gbm_bo_get_fd_for_plane(m_bo, 0); dmaBufFD >= 0) {
+        m_dmaBufFD = dmaBufFD;
+        return true;
+    }
+
+    return false;
+}
+
+void MemoryMappedGPUBuffer::storeLinearBufferInLinearFormat(uint32_t* dst, uint32_t dx, uint32_t dy, [[maybe_unused]] uint32_t dw, uint32_t dh, uint32_t dpitch,
+    const uint32_t* src, uint32_t sx, uint32_t sy, uint32_t, uint32_t sh, uint32_t spitch) const
+{
+#if CPU(ARM)
+    constexpr uint32_t numberOfPixelsPerBatch = 16;
+    constexpr uint32_t prefetchBytes = numberOfPixelsPerBatch * sizeof(uint32_t);
+    const uint32_t dwAligned = alignUpper(dw, numberOfPixelsPerBatch);
+
+    for (uint32_t y = 0; y < dh; ++y) {
+        const uint32_t* srcRow = src + (y + sy) * spitch + sx;
+        uint32_t* dstRow = dst + (y + dy) * dpitch + dx;
+
+        // Prefetch upcoming rows to reduce memory latency
+        __builtin_prefetch(srcRow + prefetchBytes, 0, 1);
+        __builtin_prefetch(dstRow + prefetchBytes, 1, 1);
+
+        uint32_t x = 0;
+        for (; x < dwAligned; x += numberOfPixelsPerBatch) {
+            // Prefetch the next memory block
+            __builtin_prefetch(srcRow + x + prefetchBytes, 0, 1);
+            __builtin_prefetch(dstRow + x + prefetchBytes, 1, 1);
+
+#if HAVE(ARM_NEON_INTRINSICS)
+            uint32x4_t v0 = vld1q_u32(srcRow + x);
+            uint32x4_t v1 = vld1q_u32(srcRow + x + 1 * sizeof(uint32_t));
+            uint32x4_t v2 = vld1q_u32(srcRow + x + 2 * sizeof(uint32_t));
+            uint32x4_t v3 = vld1q_u32(srcRow + x + 3 * sizeof(uint32_t));
```

Comment:
I may be misunderstanding the docs, but here you increase the offset by 4 every time because you're loading 4 `uint32_t` values at a time, right? So the `sizeof(uint32_t)` here (which happens to also be 4) seems misleading?

Secondly, this is a memcpy, right? I would imagine an optimized memcpy belongs in WTF/?

Thirdly, could it be that a call to memcpy might be faster here? The libc version may already be using SIMD instructions, right?